// ckeditor4-angular/esm2015/ckeditor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md.
 */
/**
 * Basic typings for the CKEditor4 elements.
 */
export var CKEditor4;
(function (CKEditor4) {
    /**
     * The CKEditor4 editor constructor.
     * @record
     */
    function Config() { }
    CKEditor4.Config = Config;
    /**
     * The event object passed to CKEditor4 event callbacks.
     *
     * See https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_eventInfo.html
     * to learn more.
     * @record
     */
    function EventInfo() { }
    CKEditor4.EventInfo = EventInfo;
    if (false) {
        /** @type {?} */
        EventInfo.prototype.name;
        /** @type {?} */
        EventInfo.prototype.editor;
        /** @type {?} */
        EventInfo.prototype.data;
        /** @type {?} */
        EventInfo.prototype.listenerData;
        /** @type {?} */
        EventInfo.prototype.sender;
        /**
         * @return {?}
         */
        EventInfo.prototype.cancel = function () { };
        /**
         * @return {?}
         */
        EventInfo.prototype.removeListener = function () { };
        /**
         * @return {?}
         */
        EventInfo.prototype.stop = function () { };
    }
})(CKEditor4 || (CKEditor4 = {}));
